plutus-core/plutus-core/src/PlutusCore/Builtin/Case.hs

```diff
@@ -19,25 +19,32 @@ import Universe
 class AnnotateCaseBuiltin uni where
     -- | Given a tag for a built-in type and a list of branches, annotate each of the branches with
     -- its expected argument types or fail if casing on values of the built-in type isn't supported.
+    -- Note: you don't need to include the resulting type of the whole case matching in the
+    -- returning list here.
     annotateCaseBuiltin
         :: UniOf term ~ uni
-        => SomeTypeIn uni
+        => ann
+        -> SomeTypeIn uni
         -> [term]
         -> Either Text [(term, [Type TyName uni ann])]
 
 class CaseBuiltin uni where
     -- | Given a constant with its type tag and a vector of branches, choose the appropriate branch
     -- or fail if the constant doesn't correspond to any of the branches (or casing on constants of
     -- this type isn't supported at all).
-    caseBuiltin :: UniOf term ~ uni => Some (ValueOf uni) -> Vector term -> Either Text term
+    caseBuiltin
+        :: UniOf term ~ uni
+        => Some (ValueOf uni)
+        -> Vector term
+        -> Either Text ([Some (ValueOf uni)], term)
```

Comment:
We should use `PlutusCore.Builtin.KnownType.HeadSpine` here except it needs to be changed to allow for the spine to contain elements of a different type than the head. What you have here isn't efficient.

Reply:
`-- Used in built-in functions returning function applications such as 'CaseList'.` alright, should've read more 😔 Thank you.